src/passes/J2CLItableMerging.cpp

```diff
@@ -0,0 +1,373 @@
+  }
+
+  // Redirects all itable access by changing `struct.get` of the `itable` field
+  // to `struct.get` on the to `vtable` field.
+  void rerouteItableAccess(Module& wasm) {
+    struct Rerouter : public WalkerPass<PostWalker<Rerouter>> {
+      bool isFunctionParallel() override { return true; }
+
+      J2CLItableMerging& parent;
+
+      Rerouter(J2CLItableMerging& parent) : parent(parent) {}
+
+      std::unique_ptr<Pass> create() override {
+        return std::make_unique<Rerouter>(parent);
+      }
+
+      void visitStructGet(StructGet* curr) {
+        if (curr->ref->type == Type::unreachable) {
+          return;
+        }
+
+        if (curr->type.isStruct() &&
+            parent.structInfoByITableType.count(curr->type.getHeapType())) {
+          // This is a struct.get that returns an itable type;
+          // Change to return the corresponding vtable type.
+          Builder builder(*getModule());
+          replaceCurrent(builder.makeStructGet(
+            0,
+            curr->ref,
+            parent.structInfoByITableType[curr->type.getHeapType()]
+              ->javaClass.getStruct()
+              .fields[0]
+              .type));
+
+          return;
+        }
+      }
+    };
+
+    Rerouter rerouter(*this);
+    rerouter.run(getPassRunner(), &wasm);
+    rerouter.runOnModuleCode(getPassRunner(), &wasm);
+  }
+
+  // Modify the struct definitions adding the itable fields to the vtable and
+  // preserving the vtable field names.
+  void updateTypes(Module& wasm) {
+    class TypeRewriter : public GlobalTypeRewriter {
+      J2CLItableMerging& parent;
+
+    public:
+      TypeRewriter(Module& wasm, J2CLItableMerging& parent)
+        : GlobalTypeRewriter(wasm), parent(parent) {}
+
+      void modifyStruct(HeapType oldStructType, Struct& struct_) override {
+        if (parent.structInfoByVtableType.count(oldStructType)) {
+          auto& newFields = struct_.fields;
+
+          auto structInfo = parent.structInfoByVtableType[oldStructType];
+          // Add the itable fields to the beginning of the vtable.
+          auto it = structInfo->itable.getStruct().fields.rbegin();
+          while (it != structInfo->itable.getStruct().fields.rend()) {
+            newFields.insert(newFields.begin(), *it++);
+            newFields[0].type = getTempType(newFields[0].type);
+          }
+
+          // Update field names as well. The Type Rewriter cannot do this for
+          // us, as it does not know which old fields map to which new ones
+          // (it just keeps the names in sequence).
+          auto iter = wasm.typeNames.find(oldStructType);
+          if (iter != wasm.typeNames.end()) {
+            auto& nameInfo = iter->second;
+
+            // Make a copy of the old ones to base ourselves off of as we do
+            auto oldFieldNames = nameInfo.fieldNames;
+
+            // Clear the old names and write the new ones.
+            nameInfo.fieldNames.clear();
```

Comment:
can't insert empty value to initial vector?

Reply:
Its not a vector but a map.